### ISSUE TITLE
How to use cudnn frontend in Jax?

### ISSUE THREAD

**Issue by MoFHeka** [NVIDIA/cudnn-frontend#89]
There is a XLA rewriter in Tensorflow with cudnn frontend. But what about Jax?

**Comment by nouiz** [comment]
TF and JAX use the same XLA code. So both sometimes use cudnn frontend.
What are you looking for here?

**Comment by MoFHeka** [comment]
It seems that cudnn fused MHA rewriter was not implemented in Jax XLA backend?在 2024年7月16日，05:13，Frédéric Bastien ***@***.***> 写道：﻿
TF and JAX use the same XLA code. So both sometimes use cudnn frontend.
What are you looking for here?

**Comment by nouiz** [comment]
To our knowledge it is the same. Can you give a repro of the issue?
Can you make sure the XLA graph in both case before optimization is the same?
Small difference can disable the optimization.
The optimization will never be as reliable as we want.
So we created a JAX API that will guarantee to use it and it just got merged recently:

https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention

Does that cover what you need?

**Comment by MoFHeka** [comment]
When I output the HLO IR from a simple multi-head attention layer, I found there is no FMHA-BHm-scale-bias-softmax-Bmm primitive. Could you tell me where is cudnn fmha rewriter pass is merged into OpenXla or other Jax ecosystem?
